### ISSUE TITLE
About javascript

### ISSUE THREAD

**Issue by zrose584** [user234683/youtube-local#10]
I really like this project.
But I think a little javascript could improve "QoL"..
I think of the following:
  * ability to jump to timestamps (e.g. 14:31)
  * lazy loading of comments (-> faster `/watch` page load)
  * load more comments while staying on the `/watch` site.
  * hotkeys (e.g. 'k' for pause [etc.](https://support.google.com/youtube/answer/7631406?hl=en))

In case you want to preserve "Works without Javascript" as a feature, all this could be made optionally.
What do you think?

**Comment by user234683** [comment]
Yes there is already a small amount of javascript sprinkled in for e.g. the ?t feature, and I'm not opposed to small amounts of it, so long as it's not excessive and is strictly an enhancement.

> ability to jump to timestamps (e.g. 14:31)

This is something in the back of my mind I've been planning on adding. Since it's pretty easy I'll go ahead and get it done soon.

> lazy loading of comments (-> faster /watch page load)

The speed improvement from this will likely be negligible, since the requests for the watch page and comments get sent out in parallel (latency is the primary bottleneck of Tor, and in the console window the comments are usually ready very quickly anyways). Lazy loading comments is also perceptually slower

> load more comments while staying on the /watch site.

I'm less enthused about this idea, though I wouldn't rule it out. This might be because I'm a heavy tab user and I always just ctrl-click to open more comments in a new tab, so I do tend to forget that most people stay on a single tab. I can see a few different solutions here:
- Change the "More comments" button on the watch page so it opens in a new tab by default
- "More comments" button sticks new comments on the end of the page, so that the comments continue to pile up
- "More comments" button replaces the current comments with those from the next page, preferably with a different button to go back to the previous page.
- All three options possible through a setting
I want to keep code complexity down so I only want to do this if a lot of people want it. I can start a new issue to get people's thoughts if you're interested in pursuing it further.

> hotkeys (e.g. 'k' for pause etc.)

This would probably be fine, I would suggest opening an issue for it to keep track of it so I (or other contributors) remember to get around to it.

**Comment by zrose584** [comment]
Its not just about new "root" comments, but also replies, right?
When loading comments in a new tab, I don't see how a "jump to timestamp" feature could still be implemented.
(New tab cant access original video-Element)

I see two ways to inject new comments.
1. render comments via javascript
2. via iframe (parent can be accessed via [window.parent](https://developer.mozilla.org/en-US/docs/Web/API/Window/parent)).


Which one do you prefer?

Another thing that came to my mind is viewing the entire

**Comment by user234683** [comment]
Yes, the jump to timestamp feature I'm planning to add would only apply to video descriptions and the first page of comments for now. I want to avoid iframes because the browser back button acts on them instead of the parent page and there's no (non-javascript) way I know of to disable it. Instead of rendering comments via javascript, I would prefer having an ajax request and the server generates the html and the javascript either appends it or replaces the current comments with it by doing `innerHTML = ...` (this would result in less code duplication than building a big javascript comment rendering system. Avoiding javascript rendering is one reason I built this in the first place). This could perhaps be done by adding a query parameter to the `/comments` route, something like `comments_only=true` or `ajax=true` or something like that which causes the resulting page to not include the little header on more comments pages such as the video title and thumbnail and to not include the youtube-local page header (search bar, playlist form ,etc). Simplest way to do that second part would probably be to just add an if statement above the Jinja `extends` statement in the template file.

**Comment by zrose584** [comment]
> by doing `innerHTML = ...`

good idea!

**Comment by user234683** [comment]
One idea would be to put the button under the video Uploaded/Published text, in the same row as the checkbox for adding the current vid to playlists. Might want to shorten the text to something like "Copy URL at current time" so it fits.

**Comment by zrose584** [comment]
What do you think about putting a \<select> for quality in that same row?
If quality > 720p (or live stream) it could load (hls.js)[https://hls-js.netlify.app/demo/] (lazily)?
One bad thing about hls.js though is, that it uses xhrRequests which include the "origin: localhost:port" header..

**Comment by user234683** [comment]
That should work (idk if it might look crowded, though as you can probably tell, aesthetics is not my forté). I wouldn't worry about the origin header, since the browser is already sending the referrer as it stands. But that is a good point that the port is being sent to googlevideo since that varies per user. Once I finish Tor (and non-Tor) video routing, I should be able to remove all such headers.

hls.js is 21.5K lines of code (600kb). That seems like quite a lot just to parse some m3u8 text files and stitch stream parts together into MediaSource. Is there a leaner alternative for our use case here or is all of that essential complexity? We could also use DASH streaming if that's simpler.
